### PR TITLE
EES-2509 Add new Caching service to cache results in Azure blob storage

### PR DESCRIPTION
**Pull request by benoutram**
This PR introduces a new caching interface with implementation `BlobStorageCacheService` which can read and write cached entity results in blob storage serialised as JSON, identified by a unique key.

It is a the same 'Cache Aside' strategy that we were already using to cache data block results where we first try to read from the cache, and if no result is found then read the result from the database, serialising the result as JSON in the storage container before returning it.

The same logic is moved to this new service so that it can also be used by the Public Content API's endpoint to retrieve the themes 'tree' of all Methodologies via the cache, significantly improving the load time of this page from around 1.5s to <100ms.

The interface of the cache service expects a callback function to provide the entity result.

In future it's expected that other use cases will need to implement a method of expiring cache items.

For now, we know the code points where methodologies can become publicly accessible or their taxonomy can change, so we can invalidate the cache item by deleting the blob at these points.

We also leave expiring data blocks to any mechanism for removing them that might already exist, for example regenerating all content via the Publisher.

### CODE REVIEW

**Review by markhiveit — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Content.Api/Services/MethodologyService.cs [hunk 1]
```diff
@@ -66,37 +72,44 @@ public class MethodologyService : IMethodologyService
...
-                .Select(theme => new AllMethodologiesThemeViewModel
-                {
-                    Id = theme.Id,
-                    Title = theme.Title,
-                    Topics = theme.Topics.Select(topic => new AllMethodologiesTopicViewModel
+            return await _cacheService.GetItem(
+                PublicContent,
+                AllMethodologiesCacheKey.Instance,
+                async () =>
+            {
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Services/MethodologyService.cs:79]: Thanks I've put name identifiers on this now to make it a bit clearer and checked the formatter. I don't think there were actually any formatting differences before, but hopefully the name identifiers make it a bit clearer.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/ViewModels/TableBuilderResultViewModel.cs [hunk 2]
```diff
@@ -14,4 +18,30 @@ public TableBuilderResultViewModel()
             Results = new List<ObservationViewModel>();
         }
     }
-}
\ No newline at end of file
+
+    public class TableBuilderResultCacheKey : ICacheKey
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/ViewModels/TableBuilderResultViewModel.cs:22]: Thanks, the current naming is actually related to one of your other suggestions, that ICacheKey could have a type parameter specifying the corresponding type of data that it's associated with. I originally did this so this is the cache key for the `TableBuilderResultViewModel`, hence the name. Now that I don't have this (which I'll leave to explaining on the other comment), I'll rename this to `DataBlockTableResultCacheKey` like you say.

File: src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Extensions/MockCacheServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        /// <summary>
+        /// Set up a call on a mocked cache service that always returns the result of invoking the entity provider
+        /// </summary>
+        /// <param name="cacheService"></param>
+        /// <param name="expectedBlobContainer"></param>
+        /// <param name="expectedCacheKey"></param>
+        /// <typeparam name="TEntity"></typeparam>
+        /// <returns></returns>
+        public static Mock<ICacheService> SetupEntityProviderResult<TEntity>(
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Extensions/MockCacheServiceExtensions.cs:22]: Thanks, I struggled with the naming of this. I'll use your suggestion and update it.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/Interfaces/ICacheKey.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿#nullable enable
+namespace GovUk.Education.ExploreEducationStatistics.Common.Cache.Interfaces
+{
+    public interface ICacheKey
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/Interfaces/ICacheKey.cs:4]: Thanks, it's definitely possible, and if you check out one of the older commits you can see I actually had it set up like this originally. It does work and it was nice to constrain the type returned by using the cache key. It also removed needing to be explicit about types on some of the cache calls. 🙂 

However, I found it too restrictive as I needed the view model type of the cache key to exist in the same namespace.

This isn't a problem for data blocks or for the Content API using the cache service to read and write using the cache. It was a problem for deleting cache items from other projects that then meant I needed to make the `AllMethodologiesCacheKey` more common.

For example in my PR, we delete using `AllMethodologiesCacheKey` for the type `AllMethodologiesThemeViewModel` from Admin and Publisher, as well as the Content API, but only Content API really needs to know about `AllMethodologiesThemeViewModel` and all the other types that depends on.

It seems like partial classes could have helped out there but they don't work across projects/assemblies.

File: src/GovUk.Education.ExploreEducationStatistics.Content.Api/Services/MethodologyService.cs [hunk 5]
```diff
@@ -66,37 +72,44 @@ public class MethodologyService : IMethodologyService
...
-            var themes = await _contentDbContext.Themes
-                .Include(theme => theme.Topics)
-                .ThenInclude(topic => topic.Publications)
-                .AsNoTracking()
-                .Select(theme => new AllMethodologiesThemeViewModel
-                {
-                    Id = theme.Id,
-                    Title = theme.Title,
-                    Topics = theme.Topics.Select(topic => new AllMethodologiesTopicViewModel
+            return await _cacheService.GetItem(
```
[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Services/MethodologyService.cs:75]: That's a good point, we might be able to do this with an interceptor. Then if the interceptor is unit tested our unit tests for services might not need to include a specific test for a cached version at all. It's going to take me a little while to try this out though.

[the remark this answers is not in the document]
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Services/MethodologyService.cs:75]: As we discussed this morning, I've added this to new ticket EES-2548.

File: src/GovUk.Education.ExploreEducationStatistics.Common/Cache/Interfaces/ICacheKey.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Common/Cache/Interfaces/ICacheKey.cs:4]: Cool, thought it'd be something like that! :smile: We'll leave it as-is then!

**Review by ntsim — APPROVED** [review]

**Review by markhiveit — APPROVED** [review]

**Pull request merged**